torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py

```diff
@@ -50,6 +51,36 @@ def __init__(
         # Iteration/step in the training loop.
         self.iter = 0
 
+    def __getstate__(self):
+        r"""
+        Returns a ``Dict[str, Any]`` which will be pickled and saved.
+        ``process_group`` and ``subgroup`` are not serializable and excluded from
+        a returned state.
+        """
+        logger.warning(
+            "NOTE: Process group and subgroup are not serializable and excluded from a saved state."
+        )
+        return {
+            slot: getattr(self, slot)
+            for slot in self.__slots__ if slot != "process_group" and slot != "subgroup"
+        }
+
+    def __setstate__(self, state):
+        r"""
+        Takes provided ``state`` and retrieves ``PostLocalSGDState``.
+        ``process_group`` and ``subgroup`` are set to default process_group and subgroup respectively.
+        Default subgroup is equivalent to the subgroup on each node.
+        """
+        self.process_group = distributed_c10d._get_default_group()
+        self.subgroup, _ = dist.new_subgroups()
+        logger.warning(
+            "NOTE: Process group will be set to a default group (i.e. the world size).\n"
+            "Subgroup will be set to `None` (a new set of intra-node subgroups will be created by `postLocalSGD_hook`)\n"
```

Comment:
It looks like this is inconsistent, subgroup is set to `new_subgroups` above?

Reply:
Yes, this seems inconsistent. `dist.new_subgroups()` will return the current subgroup as the first return value:
https://github.com/pytorch/pytorch/blob/b9d516138b01adf1be6e7eb24e089df4065de5d2/torch/distributed/distributed_c10d.py#L3093-L3114
That first return value should only be `None` if no new groups are constructed.

Reply:
Thanks for spotting this, I changed the logic of restoring subgroups, but forgot to change the warning message.